File: webdriver/branches/jsapi/remote/server/src/web/tests/elementfinding_test.js
/**
 * @fileoverview Implements the tests in
 * org.openqa.selenium.ElementFindingTest using the JS API.  This file
 * should be loaded by the test_suite.js test bootstrap.
 * @author jmleyba@gmail.com (Jason Leyba)
 */

function testReturnsTitleOfPageIfSet(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  assertThat(driver.getTitle(), equals('XHTML Test Page'));
  driver.get(TEST_PAGES.simpleTestPage);
  assertThat(driver.getTitle(), equals('Hello WebDriver'));
}


function testAbleToClickOnLinkIdentifiedByText(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElement({linkText: 'click me'}).click();
  assertThat(driver.getTitle(), equals('We Arrive Here'));
}


function testDriverCanFindElementsAfterLoadingMultiplePagesInARow(driver) {
  driver.get(TEST_PAGES.formPage);
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElement({linkText: 'click me'}).click();
  assertThat(driver.getTitle(), equals('We Arrive Here'));
}


function testCanClickOnLinkIdentifiedById(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElement({id: 'linkId'}).click();
  assertThat(driver.getTitle(), equals('We Arrive Here'));
}


function testCannotFindNonExistantLinkUsingLinkText(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  assertThat('Should not have found an element by nonexistant link text',
      driver.isElementPresent({linkText: 'Not here either'}),
      is(false));
  driver.findElement({linkText: 'Not here either'});
  driver.expectErrorFromPreviousCommand();
}


function testShouldFindAnElementBasedOnId(driver) {
  driver.get(TEST_PAGES.formPage);
  var selected = driver.findElement({id: 'checky'}).isSelected();
  assertThat(selected, is(false));
}


function testCannotFindNonExistantId(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  assertThat('Should not have found an element by nonexistant ID',
      driver.isElementPresent({id: 'notThere'}),
      is(false));
  driver.findElement({id: 'notThere'});
  driver.expectErrorFromPreviousCommand();
}


function testCanFindChildrenOfANode(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElements({xpath: '/html/head'});
  driver.callFunction(function(response) {
    var elements = response.value;
    var head = elements[0];
    head.findElements({tagName: 'script'});
    driver.callFunction(function(response) {
      assertEquals(2, response.value.length);
    });
  });
}


function testReturnAnEmptyListWhenThereAreNoChildrenOfANode(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElement({id: 'table'}).
      findElements({tagName: 'tr'});
  driver.callFunction(function(response) {
    assertEquals(0, response.value.length);
  });
}


function testShouldFindElementsByName(driver) {
  driver.get(TEST_PAGES.formPage);
  var element = driver.findElement({name: 'checky'});
  assertThat(element.getValue(), is('furrfu'));
}


function testShouldFindElementsByClass(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  assertThat(
      driver.findElement({className: 'extraDiv'}).getText(),
      startsWith('Another div starts here.'));
}


function testShouldFindElementsByClassWhenItIsTheFirstNameAmongMany(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  assertThat(
      driver.findElement({className: 'nameA'}).getText(),
      equals('An H2 title'));
}


function testShouldFindElementsByClassWhenItIsTheLastNameAmongMany(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  assertThat(
      driver.findElement({className: 'nameC'}).getText(),
      equals('An H2 title'));
}


function testShouldFindElementsByClassWhenItIsInTheMiddleAmongMany(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  assertThat(
      driver.findElement({className: 'nameBnoise'}).getText(),
      equals('An H2 title'));
}


function testDoesNotFindElementByClassWhenRealNameIsShorterThanQuery(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  assertThat('Should not have found an element',
      driver.isElementPresent({className: 'nameB'}),
      is(false));
  driver.findElement({className: 'nameB'});
  driver.expectErrorFromPreviousCommand();
}


function testShouldBeAbleToFindMultipleElementsByXPath(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElements({xpath: '//div'});
  driver.callFunction(function(response) {
    var elements = response.value;
    assertTrue(
        'Should find more than 1 element, but found ' + elements.length,
        elements.length > 1);
  });
}


function testShouldBeAbleToFindMultipleElementsByLinkText(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElements({linkText: 'click me'});
  driver.callFunction(function(response) {
    assertEquals(2, response.value.length);
  });
}


function testShouldBeAbleToFindMultipleElementsByPartialLinkText(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElements({partialLinkText: 'ick me'});
  driver.callFunction(function(response) {
    assertEquals(2, response.value.length);
  });
}


function testFindElementByPartialLinkTextWhenTextIsBeginningOfLink(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElement({partialLinkText: 'Create a '}).getText();
  driver.callFunction(function(response) {
    assertEquals('Create a new anonymous window', response.value);
  });
}


function testFindElementByPartialLinkTextWhenTextIsInMiddleOfLink(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElement({partialLinkText: 'anonymous'}).getText();
  driver.callFunction(function(response) {
    assertEquals('Create a new anonymous window', response.value);
  });
}


function testFindElementByPartialLinkTextWhenTextIsAtEndOfLink(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElement({partialLinkText: 'anonymous window'}).getText();
  driver.callFunction(function(response) {
    assertEquals('Create a new anonymous window', response.value);
  });
}


function testIsElementPresentReturnsFalseWithNoMatchingPartialLinkText(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  var locator = {partialLinkText: 'anon window'};
  var futureResult = driver.isElementPresent(locator);
  assertThat('Should not have found an element', futureResult, is(false));
  driver.findElement(locator);
  driver.expectErrorFromPreviousCommand();
}


function testFindElementsReturnsEmptyArrayWhenNoMatchesByPartialLinkText(
    driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElements({partialLinkText: 'anon window'});
  driver.callFunction(function(response) {
    assertEquals(0, response.value.length);
  });
}


function testShouldBeAbleToFindMultipleElementsByName(driver) {
  driver.get(TEST_PAGES.formPage);
  driver.findElements({name: 'snack'});
  driver.callFunction(function(response) {
    var elements = response.value;
    assertTrue(
        'Should find more than 1 element, but found ' + elements.length,
        elements.length > 1);
  });
}


function testShouldBeAbleToFindMultipleElementsById(driver) {
  driver.get(TEST_PAGES.nestedPage);
  driver.findElements({id: '2'});
  driver.callFunction(function(response) {
    assertEquals(
        'Should have found 8 elements with id "2"', 8,
        response.value.length);
  });
}


function testShouldBeAbleToFindMultipleElementsByClassName(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  driver.findElements({className: 'nameC'});
  driver.callFunction(function(response) {
    assertTrue(
        'Should have found more than 1 element with class "nameC", but found ' +
            response.value.length,
        response.value.length > 1);
  });
}


function testWhenFindingByNameDoesNotReturnById(driver) {
  driver.get(TEST_PAGES.formPage);
  function assertValueIs(element, value) {
    assertThat(element.getValue(), is(value));
  }
  assertValueIs(driver.findElement({name: 'id-name1'}), 'name');
  assertValueIs(driver.findElement({id: 'id-name1'}), 'id');
  assertValueIs(driver.findElement({name: 'id-name2'}), 'name');
  assertValueIs(driver.findElement({id: 'id-name2'}), 'id');
}


function testShouldFindGrandChildren(driver) {
  driver.get(TEST_PAGES.formPage);
  driver.findElement({id: 'nested_form'}).
      findElement({name: 'x'});
}


function testShouldNotFindElementOutsideTree(driver) {
  driver.get(TEST_PAGES.formPage);
  var login = driver.findElement({name: 'login'});
  assertThat(login.isElementPresent({name: 'x'}), is(false));
  login.findElement({name: 'x'});
  driver.expectErrorFromPreviousCommand();
}


function testShouldReturnElementsThatDoNotSupportTheNameProperty(driver) {
  driver.get(TEST_PAGES.nestedPage);
  driver.findElement({name: 'div1'});
  // We're OK if this executes without error.
}


function testShouldFindHiddenElementsByName(driver) {
  driver.get(TEST_PAGES.formPage);
  driver.findElement({name: 'hidden'});
  // We're OK if this executes without error.
}


function testShouldFindAnElementBasedOnTagName(driver) {
  driver.get(TEST_PAGES.formPage);
  driver.findElement({tagName: 'input'});
  // We're OK if this executes without error.
}


function testShouldFindElementsByTagName(driver) {
  driver.get(TEST_PAGES.formPage);
  driver.findElements({tagName: 'input'});
  driver.callFunction(function(response) {
    assertEquals(
        'Should have found 19 input elements',
        19, response.value.length);
  });
}


function testFindingByCompoundClassNameIsAnError(driver) {
  driver.get(TEST_PAGES.xhtmlTestPage);
  try {
    driver.findElement({className: 'a b'});
    fail('Compound class names are not allowed');
  } catch (expected) {
  }

  try {
    driver.findElements({className: 'a b'});
    fail('Compound class names are not allowed');
  } catch (expected) {
  }
}


function testShouldBeAbleToClickOnLinksWithNoHrefAttribute(driver) {
  driver.get(TEST_PAGES.javascriptPage);
  driver.findElement({linkText: 'No href'}).click();
  assertThat(driver.getTitle(), equals('Changed'));
}


function testShouldNotBeAbleToFindAnElementOnABlankPage(driver) {
  driver.get('about:blank');
  driver.findElement({id: 'imaginaryButton'});
  driver.expectErrorFromPreviousCommand();
  assertThat(
      driver.isElementPresent({id: 'imaginaryButton'}), is(false));
}


function testFindingALinkByXpathUsingContainsKeywordShouldWork(driver) {
  driver.get(TEST_PAGES.nestedPage);
  var element = driver.findElement({xpath: '//a[contains(.,"hello world")]'});
  assertThat(element.getText(), contains('hello world'));
}

File: branches/junit4/common/test/java/org/openqa/selenium/ElementFindingTest.java
/*
Copyright 2007-2009 WebDriver committers
Copyright 2007-2009 Google Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package org.openqa.selenium;

import java.util.List;

import org.junit.Test;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.is;
import static org.openqa.selenium.Ignore.Driver.IPHONE;
import static org.openqa.selenium.Ignore.Driver.SELENESE;

public class ElementFindingTest extends AbstractDriverTestCase {

  @Test public void shouldReturnTitleOfPageIfSet() {
    driver.get(pages.xhtmlTestPage);
    assertThat(driver.getTitle(), equalTo(("XHTML Test Page")));

    driver.get(pages.simpleTestPage);
    assertThat(driver.getTitle(), equalTo("Hello WebDriver"));
  }

  @Test public void shouldNotBeAbleToLocateASingleElementThatDoesNotExist() {
    driver.get(pages.formPage);

    try {
      driver.findElement(By.id("nonExistantButton"));
      fail("Should not have succeeded");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @Test public void shouldBeAbleToClickOnLinkIdentifiedByText() {
    driver.get(pages.xhtmlTestPage);
    driver.findElement(By.linkText("click me")).click();
    assertThat(driver.getTitle(), equalTo("We Arrive Here"));
  }

  @Test public void DriverShouldBeAbleToFindElementsAfterLoadingMoreThanOnePageAtATime() {
    driver.get(pages.formPage);
    driver.get(pages.xhtmlTestPage);
    driver.findElement(By.linkText("click me")).click();
    assertThat(driver.getTitle(), equalTo("We Arrive Here"));
  }

  @Test public void shouldBeAbleToClickOnLinkIdentifiedById() {
    driver.get(pages.xhtmlTestPage);
    driver.findElement(By.id("linkId")).click();
    assertThat(driver.getTitle(), equalTo("We Arrive Here"));
  }

  @Test public void shouldThrowAnExceptionWhenThereIsNoLinkToClickAndItIsFoundWithLinkText() {
    driver.get(pages.xhtmlTestPage);

    try {
      driver.findElement(By.linkText("Not here either"));
      fail("Should not have succeeded");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @Test public void shouldfindAnElementBasedOnId() {
    driver.get(pages.formPage);

    WebElement element = driver.findElement(By.id("checky"));

    assertThat(element.isSelected(), is(false));
  }

  @Test public void shouldNotBeAbleTofindElementsBasedOnIdIfTheElementIsNotThere() {
    driver.get(pages.formPage);

    try {
      driver.findElement(By.id("notThere"));
      fail("Should not have succeeded");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @Ignore(SELENESE)
  @Test public void shouldBeAbleToFindChildrenOfANode() {
    driver.get(pages.xhtmlTestPage);
    List<WebElement> elements = driver.findElements(By.xpath("/html/head"));
    WebElement head = elements.get(0);
    List<WebElement> importedScripts = head.findElements(By.tagName("script"));
    assertThat(importedScripts.size(), equalTo(2));
  }

  @Ignore(SELENESE)
  @Test public void ReturnAnEmptyListWhenThereAreNoChildrenOfANode() {
    driver.get(pages.xhtmlTestPage);
    WebElement table = driver.findElement(By.id("table"));
    List<WebElement> rows = table.findElements(By.tagName("tr"));

    assertThat(rows.size(), equalTo(0));
  }

  @Ignore(SELENESE)
  @Test public void shouldFindElementsByName() {
    driver.get(pages.formPage);

    WebElement element = driver.findElement(By.name("checky"));

    assertThat(element.getValue(), is("furrfu"));
  }

  @Test public void shouldFindElementsByClass() {
    driver.get(pages.xhtmlTestPage);

    WebElement element = driver.findElement(By.className("extraDiv"));
    assertTrue(element.getText().startsWith("Another div starts here."));
  }

  @Test public void shouldFindElementsByClassWhenItIsTheFirstNameAmongMany() {
    driver.get(pages.xhtmlTestPage);

    WebElement element = driver.findElement(By.className("nameA"));
    assertThat(element.getText(), equalTo("An H2 title"));
  }

  @Test public void shouldFindElementsByClassWhenItIsTheLastNameAmongMany() {
    driver.get(pages.xhtmlTestPage);

    WebElement element = driver.findElement(By.className("nameC"));
    assertThat(element.getText(), equalTo("An H2 title"));
  }

  @Test public void shouldFindElementsByClassWhenItIsInTheMiddleAmongMany() {
    driver.get(pages.xhtmlTestPage);

    WebElement element = driver.findElement(By.className("nameBnoise"));
    assertThat(element.getText(), equalTo("An H2 title"));
  }
  
  @Test public void shouldFindElementByClassWhenItsNameIsSurroundedByWhitespace() {
    driver.get(pages.xhtmlTestPage);
    
    WebElement element = driver.findElement(By.className("spaceAround"));
    assertThat(element.getText(), equalTo("Spaced out"));
  }
  
  @Ignore(SELENESE)
  @Test public void shouldFindElementsByClassWhenItsNameIsSurroundedByWhitespace() {
    driver.get(pages.xhtmlTestPage);
    
    List<WebElement> elements = driver.findElements(By.className("spaceAround"));
    assertThat(elements.size(), equalTo(1));
    assertThat(elements.get(0).getText(), equalTo("Spaced out"));
  }

  @Test public void shouldNotFindElementsByClassWhenTheNameQueriedIsShorterThanCandidateName() {
    driver.get(pages.xhtmlTestPage);

    try {
      driver.findElement(By.className("nameB"));
      fail("Should not have succeeded");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @Ignore(SELENESE)
  @Test public void shouldBeAbleToFindMultipleElementsByXPath() {
    driver.get(pages.xhtmlTestPage);

    List<WebElement> elements = driver.findElements(By.xpath("//div"));

    assertTrue(elements.size() > 1);
  }

  @Ignore(SELENESE)
  @Test public void shouldBeAbleToFindMultipleElementsByLinkText() {
    driver.get(pages.xhtmlTestPage);

    List<WebElement> elements = driver.findElements(By.linkText("click me"));

    assertTrue("Expected 2 links, got " + elements.size(), elements.size() == 2);
  }

  @Ignore(SELENESE)
  @Test public void shouldBeAbleToFindMultipleElementsByPartialLinkText() {
    driver.get(pages.xhtmlTestPage);

    List<WebElement> elements = driver.findElements(By.partialLinkText("ick me"));

    assertTrue(elements.size() == 2);
  }

  @Ignore(SELENESE)
  @Test public void shouldBeAbleToFindElementByPartialLinkText() {
    driver.get(pages.xhtmlTestPage);

    try {
      driver.findElement(By.partialLinkText("anon"));
    } catch (NoSuchElementException e) {
      fail("Expected element to be found");
    }
  }

  @Ignore(SELENESE)
  @Test public void shouldBeAbleToFindMultipleElementsByName() {
    driver.get(pages.nestedPage);

    List<WebElement> elements = driver.findElements(By.name("checky"));

    assertTrue(elements.size() > 1);
  }

  @Ignore(SELENESE)
  @Test public void shouldBeAbleToFindMultipleElementsById() {
    driver.get(pages.nestedPage);

    List<WebElement> elements = driver.findElements(By.id("2"));

    assertEquals(8, elements.size());
  }

  @Ignore(SELENESE)
  @Test public void shouldBeAbleToFindMultipleElementsByClassName() {
    driver.get(pages.xhtmlTestPage);

    List<WebElement> elements = driver.findElements(By.className("nameC"));

    assertTrue(elements.size() > 1);
  }

  // You don't want to ask why this is here
  @Test public void WhenFindingByNameShouldNotReturnById() {
    driver.get(pages.formPage);

    WebElement element = driver.findElement(By.name("id-name1"));
    assertThat(element.getValue(), is("name"));

    element = driver.findElement(By.id("id-name1"));
    assertThat(element.getValue(), is("id"));

    element = driver.findElement(By.name("id-name2"));
    assertThat(element.getValue(), is("name"));

    element = driver.findElement(By.id("id-name2"));
    assertThat(element.getValue(), is("id"));
  }

  @Ignore(SELENESE)
  @Test public void shouldFindGrandChildren() {
    driver.get(pages.formPage);
    WebElement form = driver.findElement(By.id("nested_form"));
    form.findElement(By.name("x"));
  }

  @Ignore(SELENESE)
  @Test public void shouldNotFindElementOutSideTree() {
    driver.get(pages.formPage);
    WebElement element = driver.findElement(By.name("login"));
    try {
      element.findElement(By.name("x"));
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @Ignore(SELENESE)
  @Test public void shouldReturnElementsThatDoNotSupportTheNameProperty() {
    driver.get(pages.nestedPage);

    driver.findElement(By.name("div1"));
    // If this works, we're all good
  }

  @Test public void shouldFindHiddenElementsByName() {
    driver.get(pages.formPage);

    try {
      driver.findElement(By.name("hidden"));
    } catch (NoSuchElementException e) {
      fail("Expected to be able to find hidden element");
    }
  }

  @Ignore(SELENESE)
  @Test public void shouldfindAnElementBasedOnTagName() {
    driver.get(pages.formPage);

    WebElement element = driver.findElement(By.tagName("input"));

    assertNotNull(element);
  }

  @Ignore(SELENESE)
  @Test public void shouldfindElementsBasedOnTagName() {
    driver.get(pages.formPage);

    List<WebElement> elements = driver.findElements(By.tagName("input"));

    assertNotNull(elements);
  }

  @Test public void findingByCompoundClassNameIsAnError() {
    driver.get(pages.xhtmlTestPage);

    try {
      driver.findElement(By.className("a b"));
      fail("Compound class names aren't allowed");
    } catch (IllegalLocatorException e) {
      // This is expected
    }

    try {
      driver.findElements(By.className("a b"));
      fail("Compound class names aren't allowed");
    } catch (IllegalLocatorException e) {
      // This is expected
    }
  }

  @JavascriptEnabled
  @Test public void shouldBeAbleToClickOnLinksWithNoHrefAttribute() {
    driver.get(pages.javascriptPage);

    WebElement element = driver.findElement(By.linkText("No href"));
    element.click();

    // if any exception is thrown, we won't get this far. Sanity check
    assertEquals("Changed", driver.getTitle());
  }

  @Ignore({SELENESE})
  @Test public void shouldNotBeAbleToFindAnElementOnABlankPage() {
    driver.get("about:blank");

    try {
      // Search for anything. This used to cause an IllegalStateException in IE.
      driver.findElement(By.tagName("a"));
      fail("Should not have been able to find a link");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @Ignore({IPHONE})
  @NeedsFreshDriver
  @Test public void shouldNotBeAbleToLocateASingleElementOnABlankPage() {
    // Note we're on the default start page for the browser at this point.

    try {
      driver.findElement(By.id("nonExistantButton"));
      fail("Should not have succeeded");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @JavascriptEnabled
  @Test public void RemovingAnElementDynamicallyFromTheDomShouldCauseAStaleRefException() {
    driver.get(pages.javascriptPage);

    RenderedWebElement toBeDeleted = (RenderedWebElement) driver.findElement(By.id("deleted"));
    assertTrue(toBeDeleted.isDisplayed());

    driver.findElement(By.id("delete")).click();

    try {
      toBeDeleted.isDisplayed();
      fail("Element should be stale at this point");
    } catch (StaleElementReferenceException e) {
      // this is expected
    }
  }

  @Test public void findingALinkByXpathUsingContainsKeywordShouldWork() {
    driver.get(pages.nestedPage);

    try {
      driver.findElement(By.xpath("//a[contains(.,'hello world')]"));
    } catch (Exception e) {
      e.printStackTrace();
      fail("Should not have thrown an exception");
    }
  }

  @JavascriptEnabled
  @Test public void shouldBeAbleToFindAnElementByCssSelector() {
    driver.get(pages.xhtmlTestPage);
    if (!supportsSelectorApi()) {
      System.out.println("Skipping test: selector API not supported");
      return;
    }
    driver.findElement(By.cssSelector("div.content"));
  }

  @JavascriptEnabled
  @Test public void shouldBeAbleToFindAnElementsByCssSelector() {
    driver.get(pages.xhtmlTestPage);
    if (!supportsSelectorApi()) {
      System.out.println("Skipping test: selector API not supported");
      return;
    }
    driver.findElements(By.cssSelector("p"));
  }
}
